File: src-tauri/src/env_files.rs
// ──────────────────────────────────────────────────────────
// .env file editor — Tauri commands
// ──────────────────────────────────────────────────────────
//
// Dedicated commands instead of tauri-plugin-fs because:
//   - Env files live at predictable paths relative to the engine
//     root; we can scope writes without exposing a general fs
//     surface to the webview.
//   - Parsing + serialising stays server-side so the frontend
//     doesn't need to know about comments, blank lines, quotes.
//   - A later "variable is a secret" masking decision can live
//     here without round-tripping to Rust.
//
// File format: standard KEY=VALUE per line. Lines starting with
// `#` are comments; blank lines preserved on round-trip.
// ──────────────────────────────────────────────────────────

use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

use crate::sidecar::SidecarState;

#[derive(Serialize, Deserialize, Clone)]
pub struct EnvVar {
    pub key: String,
    pub value: String,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct EnvFile {
    /// Absolute path to the file.
    pub path: String,
    /// Filename only, e.g. ".env.local".
    pub filename: String,
    /// Ordered list of KEY=VALUE pairs. Comments + blanks are
    /// preserved during round-trip via `raw_lines`, but they
    /// are NOT in this list — the frontend only edits variables.
    pub variables: Vec<EnvVar>,
    /// True when the file is covered by a .gitignore rule.
    /// Best-effort: checks for a literal line match in any
    /// .gitignore between the file and the project root.
    pub gitignored: bool,
    /// Raw original lines, used to preserve comments + ordering
    /// that aren't represented in `variables`.
    pub raw_lines: Vec<String>,
}

fn parse_env_line(line: &str) -> Option<EnvVar> {
    let trimmed = line.trim_start();
    if trimmed.is_empty() || trimmed.starts_with('#') {
        return None;
    }
    let (key, value) = trimmed.split_once('=')?;
    let key = key.trim().to_string();
    if key.is_empty() || !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
        return None;
    }
    // Strip a single pair of surrounding quotes if present; otherwise keep
    // the raw value verbatim (including trailing whitespace stripped).
    let raw = value.trim_end_matches('\r');
    let stripped = if (raw.starts_with('"') && raw.ends_with('"') && raw.len() >= 2)
        || (raw.starts_with('\'') && raw.ends_with('\'') && raw.len() >= 2)
    {
        &raw[1..raw.len() - 1]
    } else {
        raw
    };
    Some(EnvVar {
        key,
        value: stripped.to_string(),
    })
}

fn is_gitignored(file: &Path, root: &Path) -> bool {
    let filename = match file.file_name().and_then(|n| n.to_str()) {
        Some(f) => f,
        None => return false,
    };
    // Walk from the file upwards looking for .gitignore files; return true
    // on first literal match. Doesn't try to be a full gitignore parser —
    // 99% of the time `.env.local` appears verbatim.
    let mut dir = file.parent();
    while let Some(d) = dir {
        let gi = d.join(".gitignore");
        if let Ok(content) = fs::read_to_string(&gi) {
            for raw in content.lines() {
                let line = raw.trim();
                if line.is_empty() || line.starts_with('#') {
                    continue;
                }
                if line == filename
                    || line == format!(".env*")
                    || line == format!("*.env")
                    || line == format!("/{}", filename)
                {
                    return true;
                }
            }
        }
        if d == root {
            break;
        }
        dir = d.parent();
    }
    false
}

fn read_env_file(path: PathBuf, root: &Path) -> Option<EnvFile> {
    let filename = path.file_name()?.to_str()?.to_string();
    let content = fs::read_to_string(&path).ok()?;
    let raw_lines: Vec<String> = content.lines().map(|s| s.to_string()).collect();
    let variables: Vec<EnvVar> = raw_lines
        .iter()
        .filter_map(|line| parse_env_line(line))
        .collect();
    Some(EnvFile {
        path: path.to_string_lossy().into_owned(),
        filename,
        variables,
        gitignored: is_gitignored(&path, root),
        raw_lines,
    })
}

#[tauri::command]
pub fn list_env_files(state: tauri::State<'_, SidecarState>) -> Result<Vec<EnvFile>, String> {
    let root = match state.current_root() {
        Some(r) => r,
        None => return Ok(Vec::new()),
    };
    let mut out = Vec::new();
    let entries = fs::read_dir(&root).map_err(|e| format!("read_dir: {}", e))?;
    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_file() {
            continue;
        }
        let name = match path.file_name().and_then(|n| n.to_str()) {
            Some(n) => n,
            None => continue,
        };
        if !(name == ".env" || name.starts_with(".env.")) {
            continue;
        }
        if let Some(file) = read_env_file(path, &root) {
            out.push(file);
        }
    }
    out.sort_by(|a, b| a.filename.cmp(&b.filename));
    Ok(out)
}

/// Replace the variable values for the file at `path` while preserving
/// any comments and blank lines from the original. Appends new
/// variables at the end. Writes atomically via a temp-file rename.
#[tauri::command]
pub fn save_env_file(
    state: tauri::State<'_, SidecarState>,
    path: String,
    variables: Vec<EnvVar>,
) -> Result<(), String> {
    let root = state
        .current_root()
        .ok_or_else(|| "no project root".to_string())?;
    let target = PathBuf::from(&path);
    // Safety: only allow writes within the open project root.
    let canon_root = root.canonicalize().map_err(|e| e.to_string())?;
    let canon_target = target
        .parent()
        .and_then(|p| p.canonicalize().ok())
        .ok_or_else(|| "invalid path".to_string())?;
    if !canon_target.starts_with(&canon_root) {
        return Err(format!(
            "refusing to write outside project root: {:?}",
            canon_target
        ));
    }

    // Read existing lines to preserve comments + key ordering.
    let existing = fs::read_to_string(&target).unwrap_or_default();
    let lines: Vec<String> = existing.lines().map(|s| s.to_string()).collect();

    // Build a map of the desired state for quick lookup.
    let mut desired: std::collections::BTreeMap<String, String> = variables
        .iter()
        .map(|v| (v.key.clone(), v.value.clone()))
        .collect();

    let mut out_lines: Vec<String> = Vec::new();
    for line in &lines {
        match parse_env_line(line) {
            Some(var) => {
                if let Some(new_val) = desired.remove(&var.key) {
                    out_lines.push(format!("{}={}", var.key, new_val));
                }
                // If key was removed from desired map, drop the line entirely.
            }
            None => out_lines.push(line.clone()),
        }
    }

    // Any keys left in `desired` are new — append at end.
    for (k, v) in desired {
        out_lines.push(format!("{}={}", k, v));
    }

    let mut final_content = out_lines.join("\n");
    if !final_content.ends_with('\n') {
        final_content.push('\n');
    }

    // Atomic write: temp file + rename.
    let tmp = target.with_extension(format!(
        "{}.0canvas-tmp",
        target
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("env")
    ));
    fs::write(&tmp, &final_content).map_err(|e| format!("write tmp: {}", e))?;
    fs::rename(&tmp, &target).map_err(|e| format!("rename: {}", e))?;
    Ok(())
}

File: electron/ipc/commands/env-files.ts
// ──────────────────────────────────────────────────────────
// IPC commands: .env editor
// ──────────────────────────────────────────────────────────
//
// Dedicated commands (instead of a generic fs plugin) because:
//   - Env files live at predictable paths under the engine root;
//     writes stay scoped without exposing arbitrary fs.
//   - Parsing + serialising stays main-side so the renderer never
//     deals with comments, blanks, quote stripping.
//
// Format: standard KEY=VALUE per line. `#`-prefixed lines are
// comments; blank lines preserved on round-trip via raw_lines.
// ──────────────────────────────────────────────────────────

import fs from "node:fs";
import path from "node:path";
import { currentRoot } from "../../sidecar";
import type { CommandHandler } from "../router";

interface EnvVar {
  key: string;
  value: string;
}

interface EnvFilePayload {
  path: string;
  filename: string;
  variables: EnvVar[];
  gitignored: boolean;
  rawLines: string[];
}

function resolveRoot(args: Record<string, unknown>): string | null {
  const explicit = typeof args.cwd === "string" ? args.cwd.trim() : "";
  if (explicit) return explicit;
  return currentRoot();
}

/** Parse one KEY=VALUE line. Mirrors parse_env_line() in env_files.rs:
 *  skip blank + comment lines, require alphanumeric/underscore keys,
 *  strip a single pair of surrounding quotes if present. */
function parseEnvLine(line: string): EnvVar | null {
  const trimmed = line.replace(/^\s+/, "");
  if (!trimmed || trimmed.startsWith("#")) return null;
  const eq = trimmed.indexOf("=");
  if (eq < 0) return null;
  const key = trimmed.slice(0, eq).trim();
  if (!key || !/^[A-Za-z0-9_]+$/.test(key)) return null;

  const raw = trimmed.slice(eq + 1).replace(/\r$/, "");
  const quoted =
    (raw.startsWith('"') && raw.endsWith('"') && raw.length >= 2) ||
    (raw.startsWith("'") && raw.endsWith("'") && raw.length >= 2);
  const value = quoted ? raw.slice(1, -1) : raw;
  return { key, value };
}

/** Best-effort gitignore check: literal-match the filename (or
 *  common env globs) against every .gitignore between file and
 *  project root. */
function isGitignored(filePath: string, root: string): boolean {
  const filename = path.basename(filePath);
  let dir = path.dirname(filePath);
  const rootResolved = path.resolve(root);

  while (true) {
    const gi = path.join(dir, ".gitignore");
    if (fs.existsSync(gi)) {
      try {
        const content = fs.readFileSync(gi, "utf-8");
        for (const raw of content.split("\n")) {
          const line = raw.trim();
          if (!line || line.startsWith("#")) continue;
          if (
            line === filename ||
            line === ".env*" ||
            line === "*.env" ||
            line === `/${filename}`
          ) {
            return true;
          }
        }
      } catch {
        /* unreadable .gitignore — skip */
      }
    }
    if (path.resolve(dir) === rootResolved) break;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return false;
}

function readEnvFile(filePath: string, root: string): EnvFilePayload | null {
  let content: string;
  try {
    content = fs.readFileSync(filePath, "utf-8");
  } catch {
    return null;
  }
  const rawLines = content.split("\n");
  // split("\n") keeps a trailing empty string when content ends with
  // a newline; strip it so the editor sees the same logical rows.
  if (rawLines.length > 0 && rawLines[rawLines.length - 1] === "") {
    rawLines.pop();
  }
  const variables: EnvVar[] = [];
  for (const line of rawLines) {
    const v = parseEnvLine(line);
    if (v) variables.push(v);
  }
  return {
    path: filePath,
    filename: path.basename(filePath),
    variables,
    gitignored: isGitignored(filePath, root),
    rawLines,
  };
}

export const listEnvFiles: CommandHandler = (args) => {
  const root = resolveRoot(args);
  if (!root) return [];
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(root, { withFileTypes: true });
  } catch (err) {
    throw new Error(
      `read_dir: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  const out: EnvFilePayload[] = [];
  for (const entry of entries) {
    if (!entry.isFile()) continue;
    const name = entry.name;
    if (!(name === ".env" || name.startsWith(".env."))) continue;
    const abs = path.join(root, name);
    const file = readEnvFile(abs, root);
    if (file) out.push(file);
  }
  out.sort((a, b) => a.filename.localeCompare(b.filename));
  return out;
};

export const saveEnvFile: CommandHandler = (args) => {
  const root = resolveRoot(args);
  if (!root) throw new Error("no project root");
  const target = String(args.path ?? "");
  if (!target) throw new Error("save_env_file: missing path");
  const variables = Array.isArray(args.variables)
    ? (args.variables as Array<{ key?: unknown; value?: unknown }>)
    : [];

  // Containment check: canonicalize target's parent and project
  // root, refuse writes outside. Use canonicalize+
  // starts_with gate.
  const canonRoot = fs.realpathSync.native(path.resolve(root));
  const parentDir = path.dirname(target);
  let canonParent: string;
  try {
    canonParent = fs.realpathSync.native(path.resolve(parentDir));
  } catch {
    throw new Error("invalid path");
  }
  if (!canonParent.startsWith(canonRoot)) {
    throw new Error(`refusing to write outside project root: ${canonParent}`);
  }

  // Read existing file (may not exist) to preserve comments + order.
  let existing = "";
  try {
    existing = fs.readFileSync(target, "utf-8");
  } catch {
    /* doesn't exist → treat as empty */
  }
  const lines = existing.split("\n");
  if (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();

  // Map of desired state — consumed as we rewrite lines, leftovers
  // append at the end.
  const desired = new Map<string, string>();
  for (const v of variables) {
    if (typeof v.key === "string" && typeof v.value === "string") {
      desired.set(v.key, v.value);
    }
  }

  const outLines: string[] = [];
  for (const line of lines) {
    const parsed = parseEnvLine(line);
    if (parsed) {
      if (desired.has(parsed.key)) {
        const newVal = desired.get(parsed.key)!;
        outLines.push(`${parsed.key}=${newVal}`);
        desired.delete(parsed.key);
      }
      // key removed from desired → drop the line entirely
    } else {
      outLines.push(line);
    }
  }
  // New keys — append sorted so diffs stay stable turn-to-turn.
  const leftover = [...desired.entries()].sort(([a], [b]) => a.localeCompare(b));
  for (const [k, v] of leftover) outLines.push(`${k}=${v}`);

  let finalContent = outLines.join("\n");
  if (!finalContent.endsWith("\n")) finalContent += "\n";

  // Atomic write: temp file + rename.
  const ext = path.extname(target).replace(/^\./, "") || "env";
  const tmp = target.replace(/\.[^.]+$/, "") + `.${ext}.zeros-tmp`;
  fs.writeFileSync(tmp, finalContent);
  fs.renameSync(tmp, target);
};
